### PR TITLE
RS NF4: support encoding/decoding using packets

### PR DESCRIPTION
**Pull request by lamphamsy**
RS over gf::NF4 can encode/decode using packets now

### CODE REVIEW

File: src/fec_rs_nf4.h [hunk 1]
```diff
@@ -251,6 +260,101 @@ class RsNf4 : public FecCode<T> {
...
+            const int frag_id = fragments_ids.get(i);
+            T* chunk = words->get(i);
+
+            // the vector will contain marked symbols that will be packed
+            // firstly. Since locations are stored in unordered map, the vector
+            // will be sorted later to facilitate packing un-marked symbols
+            std::vector<size_t> packed_symbs;
+            // pack marked symbols
+            for (auto const& data : props[frag_id].get_map()) {
+                off_t loc_offset = data.first.get_offset();
```
Comment (slaperche-scality) [inline src/fec_rs_nf4.h]: const

File: src/gf_ring.h [hunk 2]
```diff
@@ -379,7 +379,7 @@ void RingModN<T>::mul_coef_to_buf(T a, T* src, T* dest, size_t len) const
     DoubleSizeVal<T> coef = DoubleSizeVal<T>(a);
     for (i = 0; i < len; i++) {
         // perform multiplication
-        dest[i] = T((coef * src[i]) % this->_card);
+        dest[i] = mul(coef, src[i]);
```
[the remark this answers is not in the document]
Comment (lamphamsy) [reply src/gf_ring.h]: This kind of changes would be in a separated PR.

File: src/simd_nf4.h [hunk 3]
```diff
@@ -74,25 +74,27 @@ inline aint128 expand32(uint32_t* arr, int n)
...
-    ai[1] = _mm_extract_epi32(_a, 1);
-    ai[2] = _mm_extract_epi32(_a, 2);
-    ai[3] = _mm_extract_epi32(_a, 3);
-    for (i = 0; i < n; i++) {
-        if (ai[i] == 65536)
-            flag |= (1 << i);
-        else
-            bi[i] = (aint16)ai[i];
-    }
+    ai[0] = _mm_extract_epi16(_a, 0);
```
[the remark this answers is not in the document]
Comment (lamphamsy) [reply src/simd_nf4.h]: _mm_extract_epi16 does not accept the second argument as a variable :(

File: test/ec_driver.cpp [hunk 4]
```diff
@@ -360,7 +360,8 @@ template <typename T>
 void run_fec_rs_nf4(int word_size, int n_data, int n_parities, int rflag)
 {
     quad::fec::RsNf4<T>* fec;
-    fec = new quad::fec::RsNf4<T>(word_size, n_data, n_parities);
+    size_t pkt_size = 1024;
+    fec = new quad::fec::RsNf4<T>(word_size, n_data, n_parities, pkt_size);
```
[the remark this answers is not in the document]
Comment (lamphamsy) [reply test/ec_driver.cpp]: This kind of enhancement will be in a separated PR

File: benchmark/benchmark.cpp [hunk 5]
```diff
@@ -171,7 +171,7 @@ int Benchmark<T>::init()
         fec = new quad::fec::RsGfpFft<T>(word_size, k, m);
         break;
     case EC_TYPE_RS_NF4:
-        fec = new quad::fec::RsNf4<T>(word_size, k, m);
+        fec = new quad::fec::RsNf4<T>(word_size, k, m, pkt_size);
```
[the remark this answers is not in the document]
Comment (lamphamsy) [reply benchmark/benchmark.cpp]: This kind of enhancement will be in a separated PR

**Comment by lamphamsy** [comment]
@slaperche-scality : it's ready for next reviews. Thanks :)

File: src/gf_ring.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (slaperche-scality) [reply src/gf_ring.h]: Yeah, you don't need to handle it here, I'll do it in #208

File: src/simd_nf4.h [hunk 6]
```diff
@@ -74,25 +74,27 @@ inline aint128 expand32(uint32_t* arr, int n)
...
+    ai[1] = _mm_extract_epi16(_a, 1);
+    ai[2] = _mm_extract_epi16(_a, 2);
+    ai[3] = _mm_extract_epi16(_a, 3);
+    ai[4] = _mm_extract_epi16(_a, 4);
+    ai[5] = _mm_extract_epi16(_a, 5);
+    ai[6] = _mm_extract_epi16(_a, 6);
+    ai[7] = _mm_extract_epi16(_a, 7);
+
+    flag = ai[1];
+    flag += (ai[3] > 0) ? 2 : 0;
```
[the remark this answers is not in the document]
Comment (slaperche-scality) [reply src/simd_nf4.h]: You forgot this one (but did it for `inline GroupedValues<__uint128_t> unpack(__uint128_t a, int n)`

**Pull request merged**